Fix PG::ExclusionViolation in concurrent updates

Riddlerrr opened the pull request:
This PR addresses concurrency issues where simultaneous updates to the same record could trigger a PG::ExclusionViolation due to overlapping validity ranges.

## Key Changes & Technical Details:

Fix `PG::ExclusionViolation`:

Added explicit row locking (`PERFORM 1 ... FOR UPDATE`) in the UPDATE trigger. This serializes concurrent updates to the same row, preventing them from attempting to insert overlapping history ranges simultaneously.

After adding the lock we have the second issue with `RangeError`, we can fix it with slightly updated squashing logic.

Update Squashing Logic:
I updated the query that detects multiple updates for the same transaction and changed conditions from `lower(validity) = _now` to `lower(validity) >= _now` and added `upper_inf(validity)` for find only open-ended ranges. We can have only one open range by system design.

This means that when the `validity` is already updated in a parallel transaction that runs at the same millisecond, we will squash the changes and not create a new historical record.

Riddlerrr committed:
Fix concurrency issues with update trigger

tagliala commented:
Any chance to run this spec in isolation?

```
$ be rspec spec/chrono_model/time_machine/concurrency_spec.rb 
Testing against Active Record 8.1.1 with Arel 10.0.0
2025-11-26 12:08:44.882 CET [46819] LOG:  checkpoint starting: immediate force wait
2025-11-26 12:08:44.887 CET [46819] LOG:  checkpoint complete: wrote 25 buffers (0.2%); 0 WAL file(s) added, 1 removed, 0 recycled; write=0.003 s, sync=0.001 s, total=0.006 s; sync files=11, longest=0.001 s, average=0.001 s; distance=5208 kB, estimate=5208 kB; lsn=9C/3B3B8888, redo lsn=9C/3B3B8850

Randomized with seed 32877

Concurrency
  concurrent deletions
    handles concurrent deletions without error (FAILED - 1)
  concurrent updates
    handles concurrent updates without PG::ExclusionViolation (FAILED - 2)

Failures:

  1) Concurrency concurrent deletions handles concurrent deletions without error
     Failure/Error: countries = Array.new(threads_count) { |i| Country.create!(name: "test-delete-#{i}") }
     
     NameError:
       uninitialized constant Country
     # ./spec/chrono_model/time_machine/concurrency_spec.rb:49:in 'block (4 levels) in <top (required)>'
     # ./spec/chrono_model/time_machine/concurrency_spec.rb:49:in 'Array#initialize'
     # ./spec/chrono_model/time_machine/concurrency_spec.rb:49:in 'Array.new'
     # ./spec/chrono_model/time_machine/concurrency_spec.rb:49:in 'block (3 levels) in <top (required)>'
     # /~.rvm/gems/ruby-3.4.7/gems/aruba-2.3.2/lib/aruba/rspec.rb:40:in 'block (2 levels) in <top (required)>'
     # /~/.rvm/gems/ruby-3.4.7/gems/aruba-2.3.2/lib/aruba/rspec.rb:27:in 'block (2 levels) in <top (required)>'

  2) Concurrency concurrent updates handles concurrent updates without PG::ExclusionViolation
     Failure/Error: Country.create!(name: 'test')
     
     NameError:
       uninitialized constant Country
     # ./spec/chrono_model/time_machine/concurrency_spec.rb:11:in 'block (3 levels) in <top (required)>'
     # /~/.rvm/gems/ruby-3.4.7/gems/aruba-2.3.2/lib/aruba/rspec.rb:40:in 'block (2 levels) in <top (required)>'
     # /~/.rvm/gems/ruby-3.4.7/gems/aruba-2.3.2/lib/aruba/rspec.rb:27:in 'block (2 levels) in <top (required)>'

Top 2 slowest examples (0.00016 seconds, 5.9% of total time):
  Concurrency concurrent deletions handles concurrent deletions without error
    0.00011 seconds ./spec/chrono_model/time_machine/concurrency_spec.rb:47
  Concurrency concurrent updates handles concurrent updates without PG::ExclusionViolation
    0.00005 seconds ./spec/chrono_model/time_machine/concurrency_spec.rb:14

Finished in 0.00261 seconds (files took 0.69954 seconds to load)
2 examples, 2 failures

Failed examples:

rspec ./spec/chrono_model/time_machine/concurrency_spec.rb:47 # Concurrency concurrent deletions handles concurrent deletions without error
rspec ./spec/chrono_model/time_machine/concurrency_spec.rb:14 # Concurrency concurrent updates handles concurrent updates without PG::ExclusionViolation

Randomized with seed 32877

```

Riddlerrr commented:
@tagliala This is expected behavior. You should run `rake` first to setup the default rails app for the tests. After the setup you can run separate tests as usual. 
The rake command will re-create the rails app with latest DDL, so you have to run it each time you changing the DDL (template queries)

tagliala commented:
Got it, but it is not expected there to fail because specs under `chrono_model` do not require the rails application

```
 for file in spec/chrono_model/**/*_spec.rb; do be rspec $file; done
```

They all pass in isolation

Specs requiring rails are placed under `aruba/` IIRC, but it is weird to have this under aruba, because this issue is related to the chronomodel core

Riddlerrr commented:
@tagliala Thanks for clarification 👍
I'll make them isolated then

tagliala commented on README.sql:
are most of the changes in this file supposed to be backported to master?

Riddlerrr replied:
The changes not related to the `perform 1 from temporal.countries where id = old.id for update;` and `>=` is already in DDL on master. I just actualized them here to match the actual SQLs.

tagliala replied:
Can we backport the changes to master in a separate PR so we do not have unrelated changes here and we have a proper master?

Riddlerrr replied:
Sure, that makes sense 👍
I'll do it soon

Riddlerrr replied:
✅ Done

Riddlerrr replied:
Created the separate PR with readme updates: https://github.com/ifad/chronomodel/pull/395

tagliala commented on lib/chrono_model/adapter/ddl.rb:
```diff
@@ -5,7 +5,7 @@
 module ChronoModel
   class Adapter < ActiveRecord::ConnectionAdapters::PostgreSQLAdapter
-    module DDL
+    module DDL # rubocop:disable Metrics/ModuleLength
```
confirmed this change is now unnecessary

Riddlerrr replied:
✅ fixed

tagliala commented on lib/chrono_model/adapter/ddl.rb:
```diff
@@ -154,10 +154,12 @@ def chrono_create_UPDATE_trigger(table, pk, current, history, fields, values, op
                         RETURN NEW;
                     END IF;
 
+                    PERFORM 1 FROM #{current} WHERE #{pk} = OLD.#{pk} FOR UPDATE;
```
# My comment: Possibly legit ✅  

You raise an excellent point. Let me analyze this carefully.

## Analysis

You're referring to a trigger or function that uses:

```sql
perform 1 from temporal.countries where id = old.id for update;
```

And suggesting it should be:

```sql
perform 1 from only temporal.countries where id = old.id for update;
```

**Your reasoning is correct.** In PostgreSQL, when a table uses inheritance (as is the case here where `history.countries` inherits from `temporal.countries`), a query against the parent table **without** the `ONLY` keyword will also scan and lock rows in all child tables — including `history.countries`.

This means:

1. **Without `ONLY`**: `FOR UPDATE` acquires row locks on matching rows in **both** `temporal.countries` and `history.countries`, which is unnecessary and could cause contention/deadlocks.
2. **With `ONLY`**: The lock is correctly scoped to just the `temporal.countries` table, which is the intended target.

This is the same reasoning that applies to the adapter trigger — consistency matters here, and the fix is logically sound.

**Yes, I'm sure this is a valid concern.**



You're right to notice the mismatch. Let me walk through it.

## What's locked vs. what's written

- **`PERFORM 1 FROM #{current} ... FOR UPDATE`** locks the row in the **temporal** (current) table.
- All the subsequent writes (`UPDATE`, `INSERT`) target the **history** table.

The lock on the temporal table serializes concurrent triggers for the same record, which is the main goal: it prevents two triggers from interleaving their history writes. This works because every update to a given record must pass through this trigger, so the temporal row acts as a natural mutex.

## Why it's sufficient (probably)

The temporal row lock guarantees that only one trigger at a time executes the history section for a given `pk`. By the time T2 proceeds past `FOR UPDATE`, T1 has already committed (or rolled back), so T1's history writes are visible and stable. There's no concurrent write to the same history rows.

## Why a history lock could still matter

The temporal lock protects against concurrent **trigger executions**, but it does not protect against:

1. **Direct writes to the history table** that bypass the trigger (e.g., manual `INSERT`/`UPDATE` on the history schema). These wouldn't acquire the temporal row lock and could race with the trigger.

2. **Foreign key or constraint checks** — if another transaction is reading history rows with `FOR SHARE` or similar, the trigger's `UPDATE ... SET validity` could block or conflict.

3. **DELETE trigger** — the delete trigger also writes to the history table but does **not** acquire a `FOR UPDATE` lock on the temporal row (it deletes it instead). If a DELETE and an UPDATE execute concurrently on the same record, there's a narrow window where both could be writing to history without coordinating through the same lock.

That said, in the normal case (all writes go through the view triggers, no direct history manipulation), the temporal row lock is sufficient because it's effectively a **logical mutex per record** — only one trigger body runs at a time for a given `pk`, so the history writes are already serialized.

A `FOR UPDATE` on the history row would be redundant in that normal path, and it would add overhead for every update. The current approach is the more pragmatic choice: lock the lightest resource (one temporal row) to serialize access to the heavier resource (history rows).

**Within chronomodel's normal operation there are no realistic scenarios where a separate history row lock would matter.** Here's why:

1. **All three triggers (INSERT, UPDATE, DELETE) go through the view.** The application never writes to the history table directly — all CUD operations hit the public view, which fires the INSTEAD OF triggers. There is no code path in chronomodel that issues raw `INSERT`/`UPDATE` on `history.*` outside of these triggers.

2. **The DELETE trigger doesn't race with UPDATE.** Although the DELETE trigger doesn't acquire a `FOR UPDATE` on the temporal row (it deletes it instead), this isn't a problem: PostgreSQL's own row-level locking on the temporal table serializes this naturally. A concurrent `DELETE FROM temporal.x WHERE id = 1` and `UPDATE temporal.x WHERE id = 1` will block each other at the PostgreSQL level because both need to modify the same temporal row. The second one will either see the row gone (and do nothing) or proceed after the first commits.

3. **History `#save` is constrained.** The `HistoryModel` does allow `.save` on history records (via `with_hid_pkey`), but this is for **editing historical data** (e.g. correcting a past entry). It updates by `hid` (the history primary key), which is a different row than any trigger would be touching concurrently. There's no overlap with the validity-range manipulation done by triggers.

4. **The exclusion constraint is the ultimate safety net.** The history table has `EXCLUDE USING gist (id WITH =, validity WITH &&)` — even if something unexpected happened, PostgreSQL would reject overlapping ranges rather than silently corrupt the timeline.

So the temporal row lock introduced by this PR is sufficient for all scenarios that exist in chronomodel. A history-side lock would be defensive overkill with no concrete scenario to justify the extra overhead.

Riddlerrr replied:
It seems it's correct suggestion 👍
I applied it.

tagliala commented on spec/chrono_model/time_machine/concurrency_spec.rb:
```diff
@@ -0,0 +1,89 @@
+      threads = Array.new(threads_count) do |i|
+        Thread.new do
+          Foo.transaction do
+            foos[i].destroy
```
# My comment: Unsure 🤷🏼‍♂️ 

## Primary Sources

### 1. Rails documentation — `ActiveRecord::Core`

The Rails API docs explicitly state that **ActiveRecord objects are not thread-safe**. Individual AR instances maintain internal mutable state (attribute hashes, association caches, dirty tracking, the `@new_record` / `@destroyed` flags, etc.) with no synchronization.

### 2. Rails Guide on Threading and Code Execution

The [Threading and Code Execution guide](https://guides.rubyonrails.org/threading_and_code_execution.html) makes it clear that while **`ActiveRecord::Base.connection_pool`** is thread-safe (each thread checks out its own connection), the **model instances themselves are not**.

### 3. The source code itself

If you look at what `.destroy` does, it mutates the instance in place:

- Sets `@destroyed = true`
- Clears association caches
- Runs callbacks that read/write instance state
- Freezes the attributes hash

None of these operations are synchronized. Two threads calling methods on the same instance would race over that shared mutable state.

### 4. GitHub Issues / Confirmed by Rails core

This is a well-known, **intentional** design decision. Rails core team members have consistently stated that the thread-safety boundary is:

| Thread-safe | Not thread-safe |
|---|---|
| Connection pool | Individual AR instances |
| `Rails.cache` | Shared model objects |
| `Rails.logger` | Mutable query objects |

## Why it matters in the test

Even if the test "works" most of the time, sharing `foos[i]` across threads means:

- Thread A instantiated the object (with a connection from Thread A's checkout)
- Thread B calls `.destroy` on it, which mutates the same instance and triggers callbacks
- The internal state (`@destroyed`, frozen attributes, dirty tracking) is being written without a mutex

The correct pattern, as you said, is to pass the **ID** into each thread and load/destroy inside:

```ruby
ids = foos.map(&:id)

threads = ids.map do |id|
  Thread.new { Foo.find(id).destroy }
end

threads.each(&:join)
```

This way each thread owns its own AR instance and checks out its own database connection from the pool, which is the intended Rails concurrency model.

**In short**: there's no single "thread-safety guarantee" doc because the guarantee was never made — AR instances are plain mutable Ruby objects with no internal locking, and Rails has always treated the connection pool (not the model layer) as the thread-safety boundary.

Riddlerrr replied:
I've replaced it to `Foo.where(id: foo_ids[i]).delete_all`.
I can confirm it's successfully pass with current fix and fails if I remove the the added lock.

tagliala reviewed (changes requested):
Hello,

As per Copilot's review, I've requested a change in specs

However, there are some changes in production that I'm not sure they are valid, but it is important to raise the concern

Please do not use AI to summarize my comments, it's already an AI-generated summary of the issue 😅

tagliala commented:
Commit message for this change:

```
Fix squash query to handle concurrent updates

Widen the validity lookup from `= _now` to `>= _now` so that the
squash logic finds history rows written by a concurrent transaction
whose `lower(validity)` may be slightly ahead of the current
transaction's start time.

After the new `FOR UPDATE` lock serializes concurrent updates, the
second transaction's `_now` is set before it acquires the lock and
can be earlier than the `lower(validity)` written by the first
transaction. The strict equality missed that row, falling through
to the `ELSE` branch and triggering a `PG::ExclusionViolation` or
`RangeError` on overlapping validity ranges.
```

tagliala commented on spec/chrono_model/time_machine/transaction_timestamp_spec.rb:
```diff
@@ -0,0 +1,145 @@
+RSpec.describe ChronoModel::TimeMachine do
+  include ChronoTest::TimeMachine::Helpers
+
+  after do
```
# ✅  legit suggestion

We do cleanup of history in other specs as well with `.history.delete_all`

Riddlerrr committed:
Lock for update using ONLY

Riddlerrr commented:
@tagliala Hi!
I have update the PR and applied all suggestions so it is now should be ready to merge.
Also, asked Copilot to review one more time to ensure nothing is missing.

Can you review it again?

tagliala commented:
> @tagliala Hi! I have update the PR and applied all suggestions so it is now should be ready to merge. Also, asked Copilot to review one more time to ensure nothing is missing.
> 
> Can you review it again?

https://github.com/ifad/chronomodel/commit/1eff773f41b953163fd91d63f1d5a70f9d883fde

tagliala commented on lib/chrono_model/adapter/ddl.rb:
```diff
@@ -154,10 +154,12 @@ def chrono_create_UPDATE_trigger(table, pk, current, history, fields, values, op
                     _hid := NULL;
 
-                    #{"SELECT hid INTO _hid FROM #{history} WHERE #{pk} = OLD.#{pk} AND lower(validity) = _now;" unless ENV['CHRONOMODEL_NO_SQUASH']}
+                    #{"SELECT hid INTO _hid FROM #{history} WHERE #{pk} = OLD.#{pk} AND upper_inf(validity) AND lower(validity) >= _now LIMIT 1;" unless ENV['CHRONOMODEL_NO_SQUASH']}
```
where does `upper_inf(validity)` come from? It is not even mentioned in the commit message

Riddlerrr replied:
It is postgres function which finds open-ended ranges

tagliala replied:
### GPT 5.4

upper_inf(validity) means: “is this validity range still open-ended?” For ChronoModel, that is the history row whose validity ends at infinity, so it represents the current version of the record.

A simple example:

    closed history row: validity = [10:00, 10:05), name = orig
    open history row:   validity = [10:05, infinity), name = t1

For the first row, upper_inf(validity) is false, because it already has an end time. For the second row, it is true, because it is still open.

Why that is useful in the update trigger at ddl.rb: the squash lookup is trying to reuse a history row that was already created by a concurrent update. It should only ever reuse the still-open history row, never an older row that has already been closed.

The concurrency example is:

    B commits: [10:00, 10:02) orig, [10:02, infinity) t1
    C commits: [10:02, 10:03) t1,   [10:03, infinity) t2
    A started earlier at 10:01, but commits last

If A searches only with lower(validity) >= 10:01, then both later rows match:
- [10:02, 10:03) t1
- [10:03, infinity) t2

That is too broad. The upper_inf(validity) check throws away the closed row [10:02, 10:03) and keeps only [10:03, infinity), which is the one safe row to reuse.

Short version: upper_inf(validity) means “only consider the current open history row.” That is useful because it prevents the squash logic from touching already-closed historical rows.

With the current trigger, A does not create a new history row starting at 10:01.

What happens is:

1. A started at 10:01, so its _now is 10:01.
2. B commits and creates:
  [10:00, 10:02) = orig
  [10:02, infinity) = t1
3. C commits and changes that to:
  [10:00, 10:02) = orig
  [10:02, 10:03) = t1
  [10:03, infinity) = t2
4. A finally gets the lock and runs.
5. The squash lookup says: find an open history row whose start is after or at 10:01.
6. Because of upper_inf(validity), the only eligible row is:
  [10:03, infinity) = t2
7. The trigger updates that row’s values to A’s new values, but it does not change its validity bounds.
8. Then it updates the current table row too.

So the final timeline becomes:

[10:00, 10:02) = orig
[10:02, 10:03) = t1
[10:03, infinity) = stale

And the current table also says stale.

The important subtlety is this: A’s timestamp is still 10:01, but once the trigger decides to squash into an already-existing row, it only rewrites the row’s data, not its validity start. So you do not get a new [10:01, ... ) segment.

That is why upper_inf(validity) is useful. It makes sure the row A rewrites is the current open row. Without that check, A might rewrite the closed [10:02, 10:03) row instead, leaving:

[10:00, 10:02) = orig
[10:02, 10:03) = stale
[10:03, infinity) = t2

while the current table says stale.

That second outcome is the bad corruption case, because history-backed reads would say current is t2, while the live table says stale.

So in one sentence: with upper_inf(validity), A still “wins” in the live/current sense, but it wins by overwriting the open row starting at 10:03, not by creating a new 10:01 history segment.

Riddlerrr commented:
@tagliala The commit https://github.com/ifad/chronomodel/commit/1eff773f41b953163fd91d63f1d5a70f9d883fde has valid idea! I've added the test from the commit and changed the query.

We should select only infinity ended records and since we can have only one such record we don't need a order. Record lock we applied before will prevent from adding new records, so we are safe here.

Riddlerrr commented:
I have also updated PR's description to mention `upper_inf` usage.

tagliala commented:
Can you please also change the commit message to fully cover the changes? That would be very useful in git blame and for agents

Riddlerrr commented:
@tagliala Updated

tagliala commented on spec/chrono_model/time_machine/transaction_timestamp_spec.rb:
```diff
@@ -0,0 +1,147 @@
+  describe 'records updated in the same transaction' do
+    # This test reproduces the issue where clock_timestamp() in UPDATE triggers
+    # causes records updated within the same transaction to have different timestamps,
+    # making them invisible in as_of() queries.
```
looks very accurate. Also I'm surprised of the `clock_timestamp()` wording

tagliala commented on spec/chrono_model/time_machine/transaction_timestamp_spec.rb:
```diff
@@ -0,0 +1,147 @@
+      # Query all overlappers as of the bro's creation time
+      # With clock_timestamp() in UPDATE trigger, updated overlappers would have
+      # a later timestamp and would NOT be visible here
+      overlappers_at_bro_time = Overlapper.as_of(bro_created_at).where(overlap_bro_id: bro.id)
```
looks very accurate. Also I'm surprised of the `clock_timestamp()` wording

Riddlerrr commented on lib/chrono_model/adapter/ddl.rb:
```diff
@@ -154,10 +154,12 @@ def chrono_create_UPDATE_trigger(table, pk, current, history, fields, values, op
+                    PERFORM 1 FROM ONLY #{current} WHERE #{pk} = OLD.#{pk} FOR UPDATE;
+
                     _now := timezone('UTC', now());
                     _hid := NULL;
 
-                    #{"SELECT hid INTO _hid FROM #{history} WHERE #{pk} = OLD.#{pk} AND lower(validity) = _now;" unless ENV['CHRONOMODEL_NO_SQUASH']}
+                    #{"SELECT hid INTO _hid FROM #{history} WHERE #{pk} = OLD.#{pk} AND upper_inf(validity) AND lower(validity) >= _now LIMIT 1;" unless ENV['CHRONOMODEL_NO_SQUASH']}
 
                     IF _hid IS NOT NULL THEN
                         UPDATE #{history} SET (#{fields}) = (#{values}) WHERE hid = _hid;
```
## Some more explanations from GPT 5.5

 Think of validity history like:

 ```text
   [name, validity]
 ```

 Squash behavior

 Multiple updates in the same DB transaction collapse into one history row.

 Example: transaction starts at 10:00.

 ```ruby
   Foo.transaction do
     foo = Foo.create!(name: "a")
     foo.update!(name: "b")
     foo.update!(name: "c")
   end
 ```

 Because Postgres now() is transaction-stable, every trigger sees _now = 10:00.

 With squash enabled, history ends as:

 ```text
   c | [10:00, infinity)
 ```

 No rows for a or b. They were “squashed” into the final state.

 No-squash behavior

 With CHRONOMODEL_NO_SQUASH=1, the trigger should not reuse the existing same-transaction row.

 Same example:

 ```ruby
   Foo.transaction do
     foo = Foo.create!(name: "a")
     foo.update!(name: "b")
     foo.update!(name: "c")
   end
 ```

 No-squash tries to preserve each update instead of mutating the same row:

 ```text
   a | [10:00, 10:00)
   b | [10:00, 10:00)
   c | [10:00, infinity)
 ```

 This is weird but intentional for test setups where everything runs inside one big transaction.

 Why the review comment matters

 There is another case: older transaction commits last.

 ```text
   T1 starts at 09:59
   T2 starts at 10:00, updates foo, commits
   T1 resumes, updates foo
 ```

 After T2, history has:

 ```text
   t2 | [10:00, infinity)
 ```

 When T1 updates, its _now is still 09:59.

 If the trigger blindly does:

 ```sql
   tsrange(lower(validity), _now)
 ```

 it tries:

 ```text
   tsrange(10:00, 09:59)
 ```

 That is invalid: upper bound is before lower bound.

 So yes: the reviewer is right. Even with NO_SQUASH, the trigger still needs the “future open row” guard. It just
 must avoid squashing equal timestamps.

 Lazy rule:

 ```text
   squash enabled:    lower(validity) >= _now
   no_squash enabled: lower(validity) >  _now
 ```

 > protects future rows but does not squash same-transaction rows.

tagliala commented on spec/chrono_model/time_machine/transaction_timestamp_spec.rb:
```diff
@@ -68,9 +68,10 @@
+    # This test covers the case where an UPDATE trigger uses a non-transaction-stable
+    # timestamp (e.g., clock_timestamp()), which can cause records updated within the
+    # same transaction to end up with different validity start times and become
+    # invisible in as_of() queries.
```
The comment

> ```
> # This test covers the case where an UPDATE trigger uses a non-transaction-stable
> # timestamp (e.g., clock_timestamp())
> ```

Does not sound to me.

Asked GPT 5.4:

> As written, the comment in transaction_timestamp_spec.rb reads like ChronoModel might actually use, or let callers choose, a non-transaction-stable timestamp source. That is not true.
> 
> The implementation-generated triggers use now() for validity boundaries in INSERT, UPDATE, and DELETE, including the UPDATE path at ddl.rb. In PostgreSQL, now() and current_timestamp are transaction-stable for this purpose, which matches the documented guarantee in README.md. I also did not find a supported option to swap that out; the nearby env toggle at ddl.rb only disables squashing, not the timestamp source.
> 
> A better comment would frame this as a regression guard, not as a current ChronoModel behavior. For example:
>
> ``` 
>    # This test guards the ChronoModel invariant that updates performed in the
>    # same transaction share a single validity timestamp. ChronoModel's triggers
>    # use transaction-stable database time; if that regressed to a source such
>    # as clock_timestamp(), rows updated in one transaction could become
>    # invisible to as_of() queries at that transaction timestamp.
> ```

GPT 5.4 comment does not sound either. I've also checked history: It does not appear that at any moment in time Chronomodel library used `clock_timestamp()` or timestamps not stable during the transaction.

It is not clear to me why are we commenting with this assumption, like having unstable timestamps is an option

---

Second ask to GPT 5.4:

> **Summary**
> The spec in transaction_timestamp_spec.rb is testing one core contract: rows created or updated inside the same database transaction should belong to the same historical snapshot when read back through as_of and history.
> 
> Using the OverlapBro and Overlapper temporal models defined in structure.rb, it checks five concrete behaviors:
> 1. Multiple rows inserted in one transaction are all visible at the first row’s history start time, while rows from a later transaction are not.
> 2. Parent and child rows created in the same transaction remain historically consistent through associations.
> 3. Rows inserted together get the same validity start timestamp.
> 4. Rows created and then updated in the same transaction are still visible when querying at that transaction’s timestamp, with their updated values.
> 5. Repeated updates in one transaction collapse into a final state that is visible at the original transaction timestamp.
> 6. An update in a later transaction creates a second historical version.
> 
> **How I’d Comment It**
> I would comment it in behavior-first terms, not in terms of a specific trigger function. The current update-block comment in transaction_timestamp_spec.rb is a bit too implementation-specific because it talks about clock_timestamp as if that were the thing under test. The test does not prove which PostgreSQL function is used; it proves that temporal snapshots stay consistent across one transaction.
> 
> I would rewrite that block comment like this:
> 
> > These examples verify temporal snapshot consistency within a transaction.
> > Records created or updated in the same database transaction must remain
> > visible together when queried with as_of at that transaction’s timestamp.
> > If the trigger uses a timestamp that can change during the transaction,
> > the historical snapshot can split and rows may disappear from as_of queries.
> 
> I would also tighten the last example’s comment. Instead of saying it checks non-overlapping validity ranges, I’d say it checks that a later transaction creates a second history version, because that is what the assertion actually verifies.
> 
> If you want, I can update the comments directly in transaction_timestamp_spec.rb.

Riddlerrr replied:
It seems like these tests were unnecessary. I have simplified the tests and kept one "integrations" test

tagliala commented on lib/chrono_model/adapter/ddl.rb:
```diff
@@ -134,6 +134,7 @@ def chrono_create_UPDATE_trigger(table, pk, current, history, fields, values, op
           end
 
         journal &= columns
+        squash_operator = ENV['CHRONOMODEL_NO_SQUASH'] ? '>' : '>='
```
No, sorry, this adds a new layer of complexity that we should not deal within this PR

it is fine to fix only the original issue that we have with `=` instead of `>=` and leave this switch and the "else branch" for the future

Riddlerrr replied:
Ok, sure! I'll revert

Riddlerrr commented on lib/chrono_model/adapter/ddl.rb:
```diff
@@ -154,10 +154,12 @@ def chrono_create_UPDATE_trigger(table, pk, current, history, fields, values, op
+                    PERFORM 1 FROM ONLY #{current} WHERE #{pk} = OLD.#{pk} FOR UPDATE;
+
                     _now := timezone('UTC', now());
                     _hid := NULL;
 
-                    #{"SELECT hid INTO _hid FROM #{history} WHERE #{pk} = OLD.#{pk} AND lower(validity) = _now;" unless ENV['CHRONOMODEL_NO_SQUASH']}
+                    #{"SELECT hid INTO _hid FROM #{history} WHERE #{pk} = OLD.#{pk} AND upper_inf(validity) AND lower(validity) >= _now LIMIT 1;" unless ENV['CHRONOMODEL_NO_SQUASH']}
 
```
Agreed to ignore this for the PR

Riddlerrr committed:
Fix squash query to handle concurrent updates

Widen the validity lookup from `= _now` to `>= _now` so that the squash logic finds history rows written by a concurrent transaction whose `lower(validity)` may be slightly ahead of the current transaction's start time.
Also, added `upper_inf(validity)` filter for find only open-ended ranges. We can have only one open range by system design.

After the new `FOR UPDATE` lock serializes concurrent updates, the second transaction's `_now` is set before it acquires the lock and can be earlier than the `lower(validity)` written by the first transaction. The strict equality missed that row, falling through to the `ELSE` branch and triggering a `PG::ExclusionViolation` or
`RangeError` on overlapping validity ranges.

The `IF NOT FOUND` block for the lock needs to check to fallback when the row was deleted before the lock. In this case we should just do nothing. More details here: https://github.com/ifad/chronomodel/pull/394#issuecomment-4859619804

Riddlerrr commented on lib/chrono_model/adapter/ddl.rb:
```diff
@@ -154,10 +154,12 @@ def chrono_create_UPDATE_trigger(table, pk, current, history, fields, values, op
+                    PERFORM 1 FROM ONLY #{current} WHERE #{pk} = OLD.#{pk} FOR UPDATE;
+
                     _now := timezone('UTC', now());
                     _hid := NULL;
 
-                    #{"SELECT hid INTO _hid FROM #{history} WHERE #{pk} = OLD.#{pk} AND lower(validity) = _now;" unless ENV['CHRONOMODEL_NO_SQUASH']}
+                    #{"SELECT hid INTO _hid FROM #{history} WHERE #{pk} = OLD.#{pk} AND upper_inf(validity) AND lower(validity) >= _now LIMIT 1;" unless ENV['CHRONOMODEL_NO_SQUASH']}
 
```
`upper_inf(validity)` filter removes all records except one, since we can have only one open-ended `validity` by the system design

tagliala replied:
I think this concern is legit, but I also do not understand fully this change.

```diff
-    select hid into _hid from history.countries where id = old.id and lower(validity) >= _now;
+    select hid into _hid from history.countries where id = old.id and lower(validity) >= _now order by lower(validity) desc, hid desc limit 1;
```

Previous commit at 1eff773f41b953163fd91d63f1d5a70f9d883fde did include an order. Now we introduced the `upper_inf(validity)` I'm not aware of what this means

Riddlerrr replied:
Using `upper_inf` filter should be better because it guarantees that it will return a maximum of one record => nothing to order

tagliala commented:
@arekf please check with super-intelligence tomorrow

arekf commented:
I reviewed the final state of this PR and verified it empirically: I rebuilt the
chronomodel structure (temporal/history schemas, inheritance, exclusion
constraint, view + INSTEAD OF triggers) in a scratch PostgreSQL 17 database,
once with the master triggers and once with this PR's UPDATE trigger, and ran
multi-connection concurrency tests against both. Reproduction scripts and
instructions are attached at the bottom.

## Confirmed: the fix works

- The "older transaction commits last" scenario raises `PG::DataException`
  (*range lower bound must be less than or equal to range upper bound*) on
  master; with this PR it succeeds and leaves the current row and the open
  history row consistent. A two-connection update hammer that produces
  `ExclusionViolation`s on master runs clean with this PR.
- The `upper_inf(validity)` formulation is sound: the exclusion constraint
  (`id WITH =, validity WITH &&`) guarantees at most one open-ended row per id
  (two open ranges always overlap), so the lookup is deterministic without an
  `ORDER BY`, and `LIMIT 1` is redundant-but-harmless defense. This is the
  cleanest of the variants discussed in this thread — behaviorally equivalent
  to the `ORDER BY lower(validity) DESC ... LIMIT 1` approach from 1eff773
  whenever an open row exists (the open row always has the max `lower`), but
  it states the intent directly.
- Footnote on the earlier review discussion: the `too_many_rows` concern was
  not technically accurate — non-`STRICT` `SELECT INTO` in PL/pgSQL never
  raises on multiple rows — but the non-determinism it pointed at was real,
  and 1eff773's regression scenario (stale transaction overwriting a *closed*
  segment while the open row keeps the old value) reproduces exactly as
  described without the `upper_inf` restriction.

## Findings

### 1. The new lock doesn't check whether the row still exists — an update racing a delete permanently bricks the pk

If the row is deleted (and committed) while the UPDATE trigger waits on
`FOR UPDATE`, the `PERFORM` finds no row but the trigger proceeds anyway: the
ELSE branch inserts a fresh open-ended history row for a record that no longer
exists. Reproduced on both master and this PR — so it's pre-existing, not a
regression — but the outcome is severe: the live table ends empty while
history keeps an open "ghost" row, and **every subsequent re-insert of that id
fails with `ExclusionViolation`** until history is manually cleaned (in an
update+delete+reinsert hammer, 2999 of 3000 iterations failed after the first
ghost appeared).

Since this PR adds the perfect hook, one line closes the update side of the
race:

```sql
PERFORM 1 FROM ONLY <current> WHERE <pk> = OLD.<pk> FOR UPDATE;
IF NOT FOUND THEN
    RETURN NULL;
END IF;
```

### 2. The DELETE trigger still has the exact bug class this PR fixes for UPDATE

A stale transaction that deletes after a later-timestamped update committed
still executes `tsrange(lower(validity), _now)` with `_now < lower` — I
reproduced the inverted-range `PG::DataException` with this PR's triggers
installed. So `update!` from a stale transaction now works, but `destroy` from
the same transaction shape still crashes. Worth a follow-up issue if it stays
out of scope here: the delete trigger needs the same lock plus a decision on
how to record a stale delete.

### 3. New (rare) lock-order-inversion deadlock between UPDATE and DELETE triggers

The updated UPDATE trigger now acquires locks in the order *current row →
history row*, while the DELETE trigger still acquires *history row → current
row* (its history writes precede `DELETE FROM ONLY current`). On master both
used history → current, so this cycle couldn't form. With this PR I hit one
genuine `PG::TRDeadlockDetected` in 3000 hammer iterations with no client-side
locking. PostgreSQL resolves it by aborting one transaction after
`deadlock_timeout`, so it degrades to a retryable error rather than a hang —
but it's a new failure mode on exactly the contended workloads this PR
targets. Adding the same `PERFORM ... FOR UPDATE` at the top of the DELETE
trigger would make lock ordering uniform and eliminate the cycle (natural
companion to finding 2).

### 4. Side effect worth documenting loudly: cross-transaction squash silently erases a committed version

With `orig → t1 → t2 → stale (older transaction commits last)`, the final
history is `[orig][t1][stale]` — the committed `t2` version is gone entirely,
overwritten in place, and `stale` is recorded as current since t2's timestamp
even though it became current later. `as_of()` results for the window between
t2's commit and stale's commit change retroactively. This is a defensible
last-committer-wins resolution (any alternative needs a non-transaction-stable
timestamp), and the *breaking change* label is warranted — but for an
audit-oriented gem the README/commit message should state explicitly that
overlapping transactions can now silently drop a committed version from
history rather than erroring.

### Minor notes

- The `concurrent deletions` spec only deletes disjoint records, so it
  exercises no actual contention (findings 1–3 are exactly what it doesn't
  cover).
- The `CHRONOMODEL_NO_SQUASH` gap is already acknowledged and deferred in this
  thread — agreed that's fine for this PR.
- Existing installations only get the fix after trigger recreation via the
  VERSION-gated `ChronoModel.upgrade!`, so it needs a release to propagate.

## Bottom line

The PR solves what it set out to solve and the `upper_inf` form is the right
final shape. Before merging I'd add the one-line `IF NOT FOUND` guard
(finding 1), and either extend the same lock to the DELETE trigger here or
open a follow-up issue for findings 2–3, plus document the squash semantics
from finding 4.

---

## Reproduction scripts

Requirements: PostgreSQL (tested on 17), Ruby with the `pg` gem.

```sh
# 1. Create a scratch database with the PR's triggers
createdb chrono_pr394_test
psql -d chrono_pr394_test -f setup.sql

# 2. Scenario tests: stale-commits-last fix, update/delete deadlock, ghost row
ruby test_pr394.rb

# 3. Contention hammer: concurrent update vs delete+reinsert on the same id
ruby hammer_deadlock.rb

# To compare against master's behavior, recreate the database and edit the
# UPDATE trigger in setup.sql back to the master version before loading it:
#   - remove the "perform 1 from only temporal.countries ... for update;" line
#   - change the squash lookup back to:
#       select hid into _hid from history.countries
#       where id = old.id and lower(validity) = _now;
# Then rerun the same scripts. Expected: test 1 fails with the inverted-range
# error, and the hammer's updater also hits ExclusionViolations.
```

Both scripts assume `host: localhost, user: postgres, password: postgres,
dbname: chrono_pr394_test` — adjust the `CONFIG` hash at the top to your
environment.

Observed results, PR triggers vs master triggers:

| Scenario | master | this PR |
|---|---|---|
| Test 1: stale transaction commits last | `PG::DataException` (inverted range) | ok, current == open history row; `t2` version erased from history |
| Test 2: update vs delete with current-row lock held | deadlock only if the client locks first | `PG::TRDeadlockDetected` (also 1/3000 in the hammer from pure trigger timing) |
| Test 3: update waits on `FOR UPDATE`, row deleted meanwhile | ghost open history row | ghost open history row (unchanged) |
| Hammer: update vs delete+reinsert, 3000 iterations | 3000× `ExclusionViolation` (3 in updater), ghost row persists | 2999× `ExclusionViolation` + 1 deadlock (all in deleter), ghost row persists |

<details>
<summary><code>setup.sql</code> — minimal chronomodel structure with this PR's UPDATE trigger</summary>

```sql
-- Minimal chronomodel structure with PR #394's FINAL update trigger
create extension if not exists btree_gist;
create schema temporal;
create schema history;

create table temporal.countries (
  id         serial primary key,
  name       varchar,
  updated_at timestamptz
);

create table history.countries (
  hid         bigserial primary key,
  validity    tsrange not null,
  recorded_at timestamp not null default timezone('UTC', now())
) inherits ( temporal.countries );

alter table history.countries add constraint countries_timeline_consistency
  exclude using gist ( id with =, validity with && );

create index countries_inherit_pkey on history.countries ( id );
create index index_countries_temporal_on_validity on history.countries using gist ( validity );

create view public.countries as select * from only temporal.countries;

create or replace function public.chronomodel_countries_insert() returns trigger as $$
  begin
    if new.id is null then
      new.id := nextval('temporal.countries_id_seq');
    end if;
    insert into temporal.countries ( id, name, updated_at )
    values ( new.id, new.name, new.updated_at );
    insert into history.countries ( id, name, updated_at, validity )
    values ( new.id, new.name, new.updated_at, tsrange(timezone('UTC', now()), null) );
    return new;
  end;
$$ language plpgsql;

create trigger chronomodel_insert
  instead of insert on public.countries
  for each row execute procedure public.chronomodel_countries_insert();

-- UPDATE trigger: PR #394 final state (FOR UPDATE lock + upper_inf/>= squash)
create or replace function chronomodel_countries_update() returns trigger as $$
  declare _now timestamp;
  declare _hid integer;
  declare _old record;
  declare _new record;
  begin
    if old is not distinct from new then
      return null;
    end if;

    _old := row(old.name);
    _new := row(new.name);

    if _old is not distinct from _new then
      update only temporal.countries set ( name, updated_at ) = ( new.name, new.updated_at ) where id = old.id;
      return new;
    end if;

    perform 1 from only temporal.countries where id = old.id for update;

    _now := timezone('UTC', now());
    _hid := null;

    select hid into _hid from history.countries where id = old.id and upper_inf(validity) and lower(validity) >= _now limit 1;

    if _hid is not null then
      update history.countries set ( name, updated_at ) = ( new.name, new.updated_at ) where hid = _hid;
    else
      update history.countries set validity = tsrange(lower(validity), _now)
      where id = old.id and upper_inf(validity);

      insert into history.countries ( id, name, updated_at, validity )
      values ( old.id, new.name, new.updated_at, tsrange(_now, null) );
    end if;

    update only temporal.countries set ( name, updated_at ) = ( new.name, new.updated_at ) where id = old.id;

    return new;
  end;
$$ language plpgsql;

create trigger chronomodel_update
  instead of update on public.countries
  for each row execute procedure chronomodel_countries_update();

-- DELETE trigger: unchanged master version
create or replace function chronomodel_countries_delete() returns trigger as $$
  declare _now timestamp;
  begin
    _now := timezone('UTC', now());

    delete from history.countries
    where id = old.id and validity = tsrange(_now, null);

    update history.countries set validity = tsrange(lower(validity), _now)
    where id = old.id and upper_inf(validity);

    delete from only temporal.countries
    where id = old.id;

    return old;
  end;
$$ language plpgsql;

create trigger chronomodel_delete
  instead of delete on public.countries
  for each row execute procedure chronomodel_countries_delete();
```

</details>

<details>
<summary><code>test_pr394.rb</code> — scenario tests (stale commit, deadlock, ghost row)</summary>

```ruby
# frozen_string_literal: true

require 'pg'

CONFIG = { host: 'localhost', user: 'postgres', password: 'postgres', dbname: 'chrono_pr394_test' }.freeze

def conn = PG.connect(**CONFIG)

def timeline(c, id)
  c.exec_params('select name, validity from history.countries where id = $1 order by lower(validity)', [id])
   .map { |r| "  #{r['validity'].ljust(72)} #{r['name']}" }.join("\n")
end

puts '=' * 70
puts 'TEST 1: stale transaction commits last (the scenario the PR fixes)'
puts '=' * 70
main = conn
main.exec("insert into countries (name, updated_at) values ('orig', now())")
id = main.exec('select id from countries').first['id']

stale = conn
stale.exec('BEGIN')
stale.exec('select now()') # anchor transaction timestamp
sleep 0.05
main.exec_params('update countries set name = $1, updated_at = now() where id = $2', ['t1', id])
sleep 0.05
main.exec_params('update countries set name = $1, updated_at = now() where id = $2', ['t2', id])
sleep 0.05
begin
  stale.exec_params('update countries set name = $1, updated_at = now() where id = $2', ['stale', id])
  stale.exec('COMMIT')
  cur = main.exec_params('select name from countries where id = $1', [id]).first['name']
  open_hist = main.exec_params("select name from history.countries where id = $1 and upper_inf(validity)", [id]).first['name']
  puts "RESULT: no error. current=#{cur.inspect} open_history_row=#{open_hist.inspect} #{cur == open_hist ? '(CONSISTENT)' : '(INCONSISTENT!)'}"
  puts "NOTE: value 't2' present in history? #{main.exec_params('select 1 from history.countries where id=$1 and name=$2', [id, 't2']).ntuples == 1}"
  puts "History timeline:\n#{timeline(main, id)}"
rescue PG::Error => e
  puts "RESULT: ERROR #{e.class}: #{e.message.lines.first}"
  stale.exec('ROLLBACK')
end
stale.close

puts
puts '=' * 70
puts 'TEST 2: lock-order inversion deadlock (UPDATE trigger vs DELETE trigger)'
puts '=' * 70
# Simulates: U's trigger has executed its PERFORM ... FOR UPDATE (current row
# lock) but not yet its history UPDATE, while D's delete trigger runs.
main.exec("delete from countries")
main.exec("insert into countries (name, updated_at) values ('victim', now())")
id2 = main.exec('select id from countries').first['id']

u = conn
d = conn
u.exec('BEGIN')
# same lock the trigger's first statement takes:
u.exec_params('select 1 from only temporal.countries where id = $1 for update', [id2])

# D: delete via the view. Trigger order: lock history open row, then block on U's current-row lock.
d.send_query_params('delete from countries where id = $1', [id2])
sleep 0.3 # let D reach the blocking DELETE FROM ONLY temporal

# U: now run the actual view UPDATE -> trigger will try to lock the history open row held by D
begin
  u.exec_params('update countries set name = $1, updated_at = now() where id = $2', ['upd', id2])
  u.exec('COMMIT')
  puts 'U committed without error'
rescue PG::Error => e
  puts "U: #{e.class}: #{e.message.lines.first}"
  u.exec('ROLLBACK')
end
begin
  d.get_result&.check
  d.get_result
  puts 'D completed without error'
rescue PG::Error => e
  puts "D: #{e.class}: #{e.message.lines.first}"
end
u.close
d.close

puts
puts '=' * 70
puts 'TEST 3: ghost open history row when row is deleted while UPDATE waits'
puts '=' * 70
main.exec('delete from countries')
main.exec('delete from history.countries')
main.exec("insert into countries (name, updated_at) values ('ghost-victim', now())")
id3 = main.exec('select id from countries').first['id']

del = conn
upd = conn
del.exec('BEGIN')
del.exec_params('delete from countries where id = $1', [id3]) # holds locks, uncommitted

# UPDATE sees the row in its snapshot, blocks on PERFORM ... FOR UPDATE
upd.send_query_params('update countries set name = $1, updated_at = now() where id = $2', ['ghost', id3])
sleep 0.3
del.exec('COMMIT') # unblocks the update; FOR UPDATE now finds no row

begin
  upd.get_result&.check
  upd.get_result
  puts 'UPDATE statement completed without error'
rescue PG::Error => e
  puts "UPDATE: #{e.class}: #{e.message.lines.first}"
end

live = main.exec_params('select count(*) c from countries where id = $1', [id3]).first['c'].to_i
open_rows = main.exec_params('select name, validity from history.countries where id = $1 and upper_inf(validity)', [id3])
puts "live rows in view: #{live}"
puts "open-ended history rows for deleted record: #{open_rows.ntuples}"
open_rows.each { |r| puts "  GHOST: #{r['validity']} #{r['name']}" }
puts "History timeline:\n#{timeline(main, id3)}"

del.close
upd.close
main.close
```

</details>

<details>
<summary><code>hammer_deadlock.rb</code> — update vs delete+reinsert contention hammer</summary>

```ruby
# frozen_string_literal: true

require 'pg'

CONFIG = { host: 'localhost', user: 'postgres', password: 'postgres', dbname: 'chrono_pr394_test' }.freeze

main = PG.connect(**CONFIG)
main.exec('delete from countries')
main.exec('delete from history.countries')
main.exec("insert into countries (id, name, updated_at) values (1, 'seed', now())")

errors = Hash.new(0)
mutex = Mutex.new
iterations = 3000
stop = false

updater = Thread.new do
  c = PG.connect(**CONFIG)
  i = 0
  until stop
    i += 1
    begin
      c.exec_params('update countries set name = $1, updated_at = now() where id = 1', ["u#{i}"])
    rescue PG::Error => e
      mutex.synchronize { errors["updater: #{e.class}"] += 1 }
    end
  end
  c.close
end

deleter = Thread.new do
  c = PG.connect(**CONFIG)
  iterations.times do
    c.exec('delete from countries where id = 1')
    c.exec("insert into countries (id, name, updated_at) values (1, 'seed', now())")
  rescue PG::Error => e
    mutex.synchronize { errors["deleter: #{e.class}"] += 1 }
    # reseed if needed so the loop keeps contending
    begin
      c.exec("insert into countries (id, name, updated_at) values (1, 'seed', now()) on conflict do nothing") rescue nil
    end
  end
  c.close
end

deleter.join
stop = true
updater.join

puts errors.empty? ? 'no errors' : errors.map { |k, v| "#{k}: #{v}" }.join("\n")

open_rows = main.exec('select count(*) c from history.countries where id = 1 and upper_inf(validity)').first['c']
live = main.exec('select count(*) c from countries where id = 1').first['c']
puts "final: live=#{live} open_history_rows=#{open_rows}"
main.close
```

</details>

Riddlerrr commented:
The hypermind findings looks legit to me.
I'll update the lock with `IF NOT FOUND` in this PR.

@tagliala Will we fix DELETE trigger in a separate PR?

tagliala commented:
> I'll update the lock with IF NOT FOUND in this PR.

But is there a spec that does without "if not found", and "if not found" does fix?

Riddlerrr commented:
> But is there a spec that does without "if not found", and "if not found" does fix?

Yes! I have added a new spec for that:
``` 
it 'suppresses a stale update when the current row is deleted while waiting for the lock' do
```

tagliala commented on README.sql:
```diff
@@ -111,10 +111,12 @@ create or replace function chronomodel_countries_update() returns trigger as $$
       return new;
     end if;
 
+    perform 1 from only temporal.countries where id = old.id for update;
+
```
is this missing the "if not found"?

Riddlerrr replied:
Right! Missed this part.
✅ Added

tagliala commented on lib/chrono_model/adapter/ddl.rb:
```diff
@@ -154,10 +154,15 @@ def chrono_create_UPDATE_trigger(table, pk, current, history, fields, values, op
+                    PERFORM 1 FROM ONLY #{current} WHERE #{pk} = OLD.#{pk} FOR UPDATE;
+                    IF NOT FOUND THEN
+                        RETURN NULL;
+                    END IF;
```
can we please also add the reason behind this change to the commit message (use AI to summarize the reason behind this change from Arek's comment)

Riddlerrr replied:
✅ Added small explanation and left the link to the comment in the commit

arekf commented:
Fable approves fix for finding 1 👌

======

Status: the review comment has been addressed. The PR was force-pushed this morning into a single commit 7d2eed05 ("Fix squash query to handle concurrent updates"), and its commit message now contains exactly what tagliala asked for in the thread you linked — an explanation of the reasoning, plus a link to your verification comment:

- It explains why the squash lookup widened from lower(validity) = _now to >= _now (a concurrent transaction's lower(validity) can be ahead of this transaction's _now once the FOR UPDATE lock serializes updates, and strict equality missed that row, causing PG::ExclusionViolation/RangeError).
- It explains the upper_inf(validity) filter (only one open-ended range can exist by design).
- It explains the IF NOT FOUND THEN RETURN NULL guard (do nothing if the row was deleted before the lock was acquired) and links to your comment: https://github.com/ifad/chronomodel/pull/394#issuecomment-4859619804 — I verified that link resolves to your empirical-verification comment from July 1.

The other open item from July 2 was also fixed: the README.sql example at line 118 was missing the if not found then return null; end if; block, and it's now in the diff, matching the actual trigger code.

CI is fully green — RuboCop and all 30 spec jobs (Ruby 3.0–head × Rails 7.0–edge) pass on the new head.

One loose end: a single unresolved review thread remains — tagliala's June 30 comment on transaction_timestamp_spec.rb about the inaccurate clock_timestamp() wording. It's marked outdated (Riddlerrr already simplified those tests in response), so it likely just needs to be manually resolved on GitHub. Everything else looks ready for re-review/merge.
